File: backend/src/app/api/turns.py
"""轮次创建、进程内状态、NDJSON 订阅、取消与成功历史 API。"""

import asyncio
from collections.abc import AsyncGenerator, AsyncIterator, Awaitable, Callable, Sequence
from contextlib import suppress
from typing import Literal, Protocol, TypeGuard

from fastapi import APIRouter, HTTPException, status
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, ConfigDict, Field, field_validator
from starlette.types import Receive, Scope, Send

from app.ai.service import AiSettingsConfigurationError, AiSettingsNotFoundError
from app.ai.types import JsonValue
from app.story.service import SettlementConflictError, TurnHistoryRecord
from app.story.store import StoryDataError
from app.workflow.context import (
    ContextBuildError,
    LocationSimulationContext,
    TurnContextSnapshot,
)
from app.workflow.manager import (
    ActiveTurnRunError,
    RunInvoker,
    TurnRunAlreadyClaimedError,
    TurnRunSubscription,
)
from app.workflow.runtime import NodeKey, ProgressEvent, RunStatus, TurnRun, TurnRunSummary
from app.workflow.settings import TaskSettingConfigurationError, TaskSettingNotFoundError
from app.world.service import WorldConflictError, WorldNotFoundError


class TurnRunCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    player_intent: str = Field(min_length=1)

    @field_validator("player_intent")
    @classmethod
    def reject_blank_intent(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("玩家意图不能为空")
        # 只用 strip 判断空白，不改写玩家原文。
        return value


class TurnRunCreated(BaseModel):
    run_id: str
    status: Literal[RunStatus.PENDING]


class TurnRunStatusResponse(BaseModel):
    run_id: str
    status: RunStatus
    turn_id: int | None = None
    error: str | None = None


class TurnRoleResponse(BaseModel):
    role_id: int
    content: str
    offline: bool


class TurnEventResponse(BaseModel):
    event_id: int
    location_id: str
    event_type: str
    fact: dict[str, JsonValue]


class StateChangeResponse(BaseModel):
    role_id: int
    attribute_key: str
    operation: str
    old_value: JsonValue
    operand: JsonValue
    new_value: JsonValue
    reason: str


class TurnResponse(BaseModel):
    turn_id: int
    day: int
    time_slot: str
    player_intent: str
    roles: list[TurnRoleResponse]
    events: list[TurnEventResponse]
    state_changes: list[StateChangeResponse]


type TurnEventKind = Literal[
    "node_started",
    "node_retrying",
    "node_succeeded",
    "map_completed",
    "node_failed",
    "run_succeeded",
    "run_failed",
    "run_cancelled",
]

TURN_EVENT_KINDS: tuple[TurnEventKind, ...] = (
    "node_started",
    "node_retrying",
    "node_succeeded",
    "map_completed",
    "node_failed",
    "run_succeeded",
    "run_failed",
    "run_cancelled",
)


class TurnRunEventResponse(BaseModel):
    run_id: str
    kind: TurnEventKind
    location_id: str | None
    node: NodeKey | None
    attempt: int | None
    max_attempts: int
    retrying: bool
    completed_maps: int
    total_maps: int
    elapsed_ms: int
    error: str | None = None
    turn: TurnResponse | None = None


class NdjsonStreamingResponse(StreamingResponse):
    """让 OpenAPI 与真实流共享唯一媒体类型及事件 Schema。"""

    media_type = "application/x-ndjson"

    def __init__(
        self,
        content: AsyncIterator[bytes],
        *,
        cleanup: Callable[[], Awaitable[object]],
        status_code: int = status.HTTP_200_OK,
    ) -> None:
        # FastAPI 会检查 response class 的 status_code 默认值来生成 OpenAPI。
        super().__init__(content, status_code=status_code)
        self._cleanup = cleanup

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        """即使传输在首次 body 迭代前失败，也必须等待运行清理。"""

        try:
            await super().__call__(scope, receive, send)
        finally:
            cleanup_task = asyncio.ensure_future(self._cleanup())
            try:
                await asyncio.shield(cleanup_task)
            except asyncio.CancelledError:
                with suppress(KeyError):
                    await cleanup_task
                raise
            except KeyError:
                # 仅当新的运行已淘汰旧安全摘要时发生；旧运行此时已经清理。
                pass


class TurnContextSource(Protocol):
    def build_current_turn_snapshot(
        self,
        *,
        world_id: int,
        player_intent: str,
    ) -> Sequence[LocationSimulationContext] | TurnContextSnapshot: ...


class WorldReader(Protocol):
    def get_world(self, world_id: int) -> object: ...


class TurnHistorySource(Protocol):
    def list_history(self, *, world_id: int) -> tuple[TurnHistoryRecord, ...]: ...

    def get_turn(self, turn_id: int) -> TurnHistoryRecord: ...


class TurnRunControl(Protocol):
    async def create(
        self,
        source: Sequence[LocationSimulationContext] | TurnContextSnapshot,
        invoker: RunInvoker,
    ) -> TurnRun: ...

    async def get(self, run_id: str) -> TurnRunSummary: ...

    async def claim(self, run_id: str) -> TurnRunSubscription: ...

    async def cancel(self, run_id: str) -> TurnRunSummary: ...

    async def wait(self, run_id: str) -> TurnRunSummary: ...


type RunInvokerProvider = Callable[[], RunInvoker]


def _status_response(summary: TurnRunSummary) -> TurnRunStatusResponse:
    return TurnRunStatusResponse(
        run_id=summary.run_id,
        status=summary.status,
        turn_id=summary.turn_id,
        error=summary.error,
    )


def _turn_response(record: TurnHistoryRecord) -> TurnResponse:
    return TurnResponse(
        turn_id=record.turn_id,
        day=record.day,
        time_slot=record.time_slot,
        player_intent=record.player_intent,
        roles=[
            TurnRoleResponse(role_id=item.role_id, content=item.content, offline=item.offline)
            for item in record.roles
        ],
        events=[
            TurnEventResponse(
                event_id=item.event_id,
                location_id=item.location_id,
                event_type=item.event_type,
                fact=item.fact,
            )
            for item in record.events
        ],
        state_changes=[
            StateChangeResponse(
                role_id=item.role_id,
                attribute_key=item.attribute_key,
                operation=item.operation,
                old_value=item.old_value,
                operand=item.operand,
                new_value=item.new_value,
                reason=item.reason,
            )
            for item in record.state_changes
        ],
    )


def _is_turn_event_kind(value: str) -> TypeGuard[TurnEventKind]:
    return value in TURN_EVENT_KINDS


def _event_response(event: ProgressEvent, turn: TurnResponse | None = None) -> TurnRunEventResponse:
    if not _is_turn_event_kind(event.kind):
        # 运行时与公开协议必须同步；未知事件不能静默泄漏给客户端。
        raise RuntimeError(f"未知轮次进度事件类型: {event.kind}")
    return TurnRunEventResponse(
        run_id=event.run_id,
        kind=event.kind,
        location_id=event.location_id,
        node=event.node,
        attempt=event.attempt,
        max_attempts=event.max_attempts,
        retrying=event.retrying,
        completed_maps=event.completed_maps,
        total_maps=event.total_maps,
        elapsed_ms=event.elapsed_ms,
        error=event.error,
        turn=turn,
    )


async def stream_turn_run_events(
    manager: TurnRunControl,
    history: TurnHistorySource,
    subscription: TurnRunSubscription,
) -> AsyncGenerator[bytes]:
    """流连接拥有运行生命周期；非自然终态退出必须等待取消清理。"""

    terminal_seen = False
    try:
        while True:
            event = await subscription.receive()
            turn: TurnResponse | None = None
            if event.kind == "run_succeeded":
                summary = await manager.wait(subscription.run_id)
                if summary.turn_id is None:
                    raise RuntimeError("成功轮次缺少持久化 ID")
                turn = _turn_response(history.get_turn(summary.turn_id))
            response = _event_response(event, turn)
            yield (response.model_dump_json(exclude_none=True) + "\n").encode()
            if event.kind in {"run_succeeded", "run_failed", "run_cancelled"}:
                terminal_seen = True
                return
    finally:
        if not terminal_seen:
            # 被取消的 ASGI scope 仍要等 Manager 丢弃输出并释放活动槽。
            cleanup = asyncio.create_task(manager.cancel(subscription.run_id))
            try:
                await asyncio.shield(cleanup)
            except asyncio.CancelledError:
                await cleanup
                raise


def create_turn_router(
    world_service: WorldReader,
    context_builder: TurnContextSource,
    history_service: TurnHistorySource,
    manager: TurnRunControl,
    get_invoker: RunInvokerProvider,
) -> APIRouter:
    """路由只投影受控 DTO；快照、运行生命周期和历史事实由下层拥有。"""

    router = APIRouter(prefix="/api")

    async def create_run(world_id: int, payload: TurnRunCreate) -> TurnRunCreated:
        try:
            world_service.get_world(world_id)
            snapshot = context_builder.build_current_turn_snapshot(
                world_id=world_id,
                player_intent=payload.player_intent,
            )
            run = await manager.create(snapshot, get_invoker())
        except WorldNotFoundError as error:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(error)) from None
        except (
            ActiveTurnRunError,
            AiSettingsConfigurationError,
            AiSettingsNotFoundError,
            ContextBuildError,
            TaskSettingConfigurationError,
            TaskSettingNotFoundError,
            WorldConflictError,
        ) as error:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(error)) from None
        return TurnRunCreated(run_id=run.run_id, status=RunStatus.PENDING)

    async def get_run(run_id: str) -> TurnRunStatusResponse:
        try:
            return _status_response(await manager.get(run_id))
        except KeyError:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="轮次不存在或已过期",
            ) from None

    async def cancel_run(run_id: str) -> TurnRunStatusResponse:
        try:
            return _status_response(await manager.cancel(run_id))
        except KeyError:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="轮次不存在或已过期",
            ) from None

    async def stream_run(run_id: str) -> NdjsonStreamingResponse:
        try:
            subscription = await manager.claim(run_id)
        except KeyError:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="轮次不存在或已过期",
            ) from None
        except TurnRunAlreadyClaimedError as error:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(error)) from None
        return NdjsonStreamingResponse(
            stream_turn_run_events(manager, history_service, subscription),
            cleanup=lambda: manager.cancel(run_id),
        )

    def list_turns(world_id: int) -> list[TurnResponse]:
        try:
            world_service.get_world(world_id)
            return [
                _turn_response(item) for item in history_service.list_history(world_id=world_id)
            ]
        except WorldNotFoundError as error:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(error)) from None
        except (SettlementConflictError, StoryDataError, WorldConflictError) as error:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(error)) from None

    router.add_api_route(
        "/worlds/{world_id}/turn-runs",
        create_run,
        methods=["POST"],
        response_model=TurnRunCreated,
        status_code=status.HTTP_202_ACCEPTED,
    )
    router.add_api_route(
        "/turn-runs/{run_id}",
        get_run,
        methods=["GET"],
        response_model=TurnRunStatusResponse,
    )
    router.add_api_route(
        "/turn-runs/{run_id}/stream",
        stream_run,
        methods=["GET"],
        response_model=None,
        response_class=NdjsonStreamingResponse,
        responses={
            status.HTTP_200_OK: {
                "model": TurnRunEventResponse,
                "description": "完整 NDJSON 轮次进度事件流",
            }
        },
    )
    router.add_api_route(
        "/turn-runs/{run_id}/cancel",
        cancel_run,
        methods=["POST"],
        response_model=TurnRunStatusResponse,
    )
    router.add_api_route(
        "/worlds/{world_id}/turns",
        list_turns,
        methods=["GET"],
        response_model=list[TurnResponse],
    )
    return router

File: frontend/src/api/turns.ts
import type { components } from "./generated";
import { ApiError } from "./health";

export type TurnResponse = components["schemas"]["TurnResponse"];
export type TurnRunCreated = components["schemas"]["TurnRunCreated"];
export type TurnRunEvent = components["schemas"]["TurnRunEventResponse"];
export type TurnRunStatus = components["schemas"]["TurnRunStatusResponse"];

export interface TurnRunClient {
  create(worldId: number, playerIntent: string): Promise<TurnRunCreated>;
  stream(
    runId: string,
    signal: AbortSignal,
    onEvent: (event: TurnRunEvent) => void,
  ): Promise<void>;
  cancel(runId: string): Promise<TurnRunStatus>;
  listTurns(worldId: number, signal?: AbortSignal): Promise<TurnResponse[]>;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isFiniteNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function isPositiveInteger(value: unknown): value is number {
  return isFiniteNumber(value) && Number.isInteger(value) && value > 0;
}

function isNonNegativeInteger(value: unknown): value is number {
  return isFiniteNumber(value) && Number.isInteger(value) && value >= 0;
}

function isEventKind(
  value: unknown,
): value is components["schemas"]["TurnEventKind"] {
  return (
    value === "node_started" ||
    value === "node_retrying" ||
    value === "node_succeeded" ||
    value === "map_completed" ||
    value === "node_failed" ||
    value === "run_succeeded" ||
    value === "run_failed" ||
    value === "run_cancelled"
  );
}

function isNodeKey(value: unknown): boolean {
  return (
    value === "location_simulation" || value === "attribute_memory_analysis"
  );
}

function isRunStatus(
  value: unknown,
): value is components["schemas"]["RunStatus"] {
  return (
    value === "pending" ||
    value === "running" ||
    value === "succeeded" ||
    value === "failed" ||
    value === "cancelled"
  );
}

function isTurnResponse(value: unknown): value is TurnResponse {
  if (!isRecord(value)) return false;
  const roles = value.roles;
  const events = value.events;
  const stateChanges = value.state_changes;
  return (
    isPositiveInteger(value.turn_id) &&
    isPositiveInteger(value.day) &&
    typeof value.time_slot === "string" &&
    typeof value.player_intent === "string" &&
    Array.isArray(roles) &&
    roles.every(
      (role) =>
        isRecord(role) &&
        isPositiveInteger(role.role_id) &&
        typeof role.content === "string" &&
        typeof role.offline === "boolean",
    ) &&
    Array.isArray(events) &&
    events.every(
      (event) =>
        isRecord(event) &&
        isPositiveInteger(event.event_id) &&
        typeof event.location_id === "string" &&
        typeof event.event_type === "string" &&
        isRecord(event.fact),
    ) &&
    Array.isArray(stateChanges) &&
    stateChanges.every(
      (change) =>
        isRecord(change) &&
        isPositiveInteger(change.role_id) &&
        typeof change.attribute_key === "string" &&
        typeof change.operation === "string" &&
        "old_value" in change &&
        "operand" in change &&
        "new_value" in change &&
        typeof change.reason === "string",
    )
  );
}

function isTurnRunCreated(value: unknown): value is TurnRunCreated {
  return (
    isRecord(value) &&
    typeof value.run_id === "string" &&
    value.run_id.length > 0 &&
    value.status === "pending"
  );
}

function isTurnRunStatus(value: unknown): value is TurnRunStatus {
  if (
    !isRecord(value) ||
    typeof value.run_id !== "string" ||
    !isRunStatus(value.status)
  ) {
    return false;
  }
  return (
    (value.turn_id === undefined ||
      value.turn_id === null ||
      isPositiveInteger(value.turn_id)) &&
    (value.error === undefined ||
      value.error === null ||
      typeof value.error === "string")
  );
}

function isTurnRunEvent(value: unknown): value is TurnRunEvent {
  if (
    !isRecord(value) ||
    typeof value.run_id !== "string" ||
    value.run_id.length === 0 ||
    !isEventKind(value.kind) ||
    (value.location_id !== null && typeof value.location_id !== "string") ||
    (value.node !== null && !isNodeKey(value.node)) ||
    (value.attempt !== null && !isPositiveInteger(value.attempt)) ||
    !isPositiveInteger(value.max_attempts) ||
    typeof value.retrying !== "boolean" ||
    !isNonNegativeInteger(value.completed_maps) ||
    !isNonNegativeInteger(value.total_maps) ||
    !isNonNegativeInteger(value.elapsed_ms) ||
    (value.error !== undefined &&
      value.error !== null &&
      typeof value.error !== "string") ||
    (value.turn !== undefined &&
      value.turn !== null &&
      !isTurnResponse(value.turn))
  ) {
    return false;
  }
  if (value.kind === "run_succeeded") return isTurnResponse(value.turn);
  if (value.kind === "run_failed") return typeof value.error === "string";
  return value.turn === undefined || value.turn === null;
}

async function errorFromResponse(response: Response): Promise<ApiError> {
  let message: string | null = null;
  try {
    const payload: unknown = await response.json();
    if (isRecord(payload) && typeof payload.detail === "string") {
      const normalized = payload.detail.trim().replace(/[\r\n\t]+/gu, " ");
      message = normalized.length > 0 ? normalized.slice(0, 500) : null;
    }
  } catch {
    // 错误响应体不可解析时保留稳定的通用消息，不将传输异常伪装成业务数据。
  }
  return new ApiError(message ?? "轮次请求失败。", response.status);
}

async function requestJson<T>(
  url: string,
  init: RequestInit,
  validate: (value: unknown) => value is T,
): Promise<T> {
  const response = await fetch(url, init);
  if (!response.ok) throw await errorFromResponse(response);
  let payload: unknown;
  try {
    payload = await response.json();
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new ApiError("轮次响应格式无效。", response.status);
    }
    throw error;
  }
  if (!validate(payload))
    throw new ApiError("轮次响应格式无效。", response.status);
  return payload;
}

function consumeLine(
  line: string,
  onEvent: (event: TurnRunEvent) => void,
): boolean {
  const trimmed = line.trim();
  if (trimmed.length === 0) return false;
  let payload: unknown;
  try {
    payload = JSON.parse(trimmed);
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new ApiError("轮次进度流格式无效。", 200);
    }
    throw error;
  }
  if (!isTurnRunEvent(payload)) {
    throw new ApiError("轮次进度流格式无效。", 200);
  }
  onEvent(payload);
  return (
    payload.kind === "run_succeeded" ||
    payload.kind === "run_failed" ||
    payload.kind === "run_cancelled"
  );
}

async function streamTurnRun(
  runId: string,
  signal: AbortSignal,
  onEvent: (event: TurnRunEvent) => void,
): Promise<void> {
  const response = await fetch(
    `/api/turn-runs/${encodeURIComponent(runId)}/stream`,
    {
      headers: { Accept: "application/x-ndjson" },
      signal,
    },
  );
  if (!response.ok) throw await errorFromResponse(response);
  if (response.body === null) {
    throw new ApiError("轮次进度流格式无效。", response.status);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let pending = "";
  let terminalSeen = false;
  try {
    while (!terminalSeen) {
      const result = await reader.read();
      if (result.done) break;
      pending += decoder.decode(result.value, { stream: true });
      let boundary = pending.indexOf("\n");
      while (boundary >= 0) {
        const line = pending.slice(0, boundary);
        pending = pending.slice(boundary + 1);
        terminalSeen = consumeLine(line, onEvent);
        if (terminalSeen) break;
        boundary = pending.indexOf("\n");
      }
    }
    if (!terminalSeen) {
      pending += decoder.decode();
      if (pending.length > 0) terminalSeen = consumeLine(pending, onEvent);
    }
    if (!terminalSeen) {
      // 未收到终态时连接结束不能视为成功，否则 UI 会永久保留幽灵活动轮次。
      throw new ApiError("轮次进度流意外断开。", response.status);
    }
  } finally {
    // 终态后不再读取后续字节，避免重复终态被再次交给 UI；主动释放读取锁。
    reader.releaseLock();
  }
}

const jsonHeaders = {
  Accept: "application/json",
  "Content-Type": "application/json",
} as const;

export const turnsApi: TurnRunClient = {
  create(worldId, playerIntent) {
    return requestJson(
      `/api/worlds/${worldId}/turn-runs`,
      {
        method: "POST",
        headers: jsonHeaders,
        body: JSON.stringify({ player_intent: playerIntent }),
      },
      isTurnRunCreated,
    );
  },
  stream: streamTurnRun,
  cancel(runId) {
    return requestJson(
      `/api/turn-runs/${encodeURIComponent(runId)}/cancel`,
      { method: "POST", headers: { Accept: "application/json" } },
      isTurnRunStatus,
    );
  },
  listTurns(worldId, signal) {
    return requestJson(
      `/api/worlds/${worldId}/turns`,
      { headers: { Accept: "application/json" }, signal },
      (value): value is TurnResponse[] =>
        Array.isArray(value) && value.every(isTurnResponse),
    );
  },
};
